spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -199,256 +183,314 @@ else if (messageListener instanceof MessageListener) {
 			throw new IllegalStateException("messageListener must be 'MessageListener' "
 					+ "or 'AcknowledgingMessageListener', not " + messageListener.getClass().getName());
 		}
-		Consumer<K, V> consumer = this.consumerFactory.createConsumer();
-		ConsumerRebalanceListener rebalanceListener = new ConsumerRebalanceListener() {
-
-			@Override
-			public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
-				logger.info("partitions revoked:" + partitions);
-			}
-
-			@Override
-			public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
-				assignedPartitions = partitions;
-				logger.info("partitions assigned:" + partitions);
-			}
-
-		};
-		if (this.partitions == null) {
-			if (topicPattern != null) {
-				consumer.subscribe(topicPattern, rebalanceListener);
-			}
-			else {
-				consumer.subscribe(Arrays.asList(topics), rebalanceListener);
-			}
-		}
-		else {
-			List<TopicPartition> topicPartitions = Arrays.asList(partitions);
-			this.definedPartitions = topicPartitions;
-			consumer.assign(topicPartitions);
-		}
-		this.consumer = consumer;
 		if (getTaskExecutor() == null) {
 			setTaskExecutor(
 					new SimpleAsyncTaskExecutor(getBeanName() == null ? "kafka-" : (getBeanName() + "-kafka-")));
 		}
-		getTaskExecutor().execute(this);
+		this.listenerConsumer = new ListenerConsumer(this.listener, this.acknowledgingMessageListener,
+				this.resetStrategy, this.recentOffset);
+		getTaskExecutor().execute(this.listenerConsumer);
 	}
 
 	@Override
 	protected void doStop() {
 		if (isRunning()) {
 			setRunning(false);
-			this.consumer.wakeup();
+			this.listenerConsumer.consumer.wakeup();
 		}
 	}
 
-	@Override
-	public void run() {
-		int count = 0;
-		long last = System.currentTimeMillis();
-		long now;
-		if (isRunning() && this.definedPartitions != null) {
-			initPartitionsIfNeeded();
-		}
-		final AckMode ackMode = getAckMode();
-		while (isRunning()) {
-			try {
-				if (logger.isTraceEnabled()) {
-					logger.trace("Polling...");
+
+
+	private class ListenerConsumer implements SchedulingAwareRunnable {
+
+		private final Log logger = LogFactory.getLog(this.getClass());
+
+		private final CommitCallback callback = new CommitCallback();
+
+		private final Consumer<K, V> consumer;
+
+		private final ConcurrentMap<String, ConcurrentMap<Integer, Long>> manualOffsets = new ConcurrentHashMap<>();
+
+		private final Map<String, Map<Integer, Long>> offsets = new HashMap<String, Map<Integer, Long>>();
+
+		private final MessageListener<K, V> listener;
+
+		private final AcknowledgingMessageListener<K, V> acknowledgingMessageListener;
+
+		private final ContainerOffsetResetStrategy resetStrategy;
```

Comment:
OK. Merging with the comment into the test-case and some other simple code polishing.
Those places where my IDEA doesn't like it. :smile: